CRM/Civicaseextras/CaseDurationLog.php

```diff
@@ -0,0 +1,388 @@
+      1 => array($caseID, 'Integer'),
+      2 => array($activityID, 'Integer'),
+      3 => array($startDate, 'String'),
+    );
+    CRM_Core_DAO::executeQuery($query, $params);
+  }
+
+  /**
+   * Ends an open case log by seting end task, end date and calculating
+   * duration.
+   *
+   * @param $activityID
+   * @param $endDate
+   * @param $caseID
+   */
+  private function endLog($activityID, $endDate, $caseID) {
+    $query = "
+      UPDATE civicrm_case_duration_log
+      SET end_task = %1, end_date = %2, duration = datediff(end_date, start_date)
+      WHERE case_id = %3
+      AND end_task IS NULL
+    ";
+    $params = array(
+      1 => array($activityID, 'Integer'),
+      2 => array($endDate, 'String'),
+      3 => array($caseID, 'Integer'),
+    );
+    CRM_Core_DAO::executeQuery($query, $params);
+  }
+
+  /**
+   * Uses data of the provided case activity object to check if the case is
+   * being opened.
+   *
+   * @param CRM_Core_DAO $caseActivity
+   *   Object with the data for the case and activity.
+   *
+   * @return bool
+   *   True if the current activity is opening the case, false otherwise
+   */
+  private function wasCaseOpening($caseActivity) {
+    $changeStatusToOpen = FALSE;
+    $activityType = CRM_Utils_Array::value($caseActivity->activity_type_id, $this->activityTypes);
+
+    if ($activityType === 'Change Case Status') {
+      if ($this->extractStatusFromSubject($caseActivity->subject) === 'Opened') {
+        $changeStatusToOpen = TRUE;
+      }
+    }
+
+    if ($activityType === 'Open Case' || $changeStatusToOpen) {
+      $query = "
+        SELECT *
+        FROM civicrm_case_duration_log
+        WHERE case_id = %1
+        AND end_task IS NULL
+      ";
+      $dbResult = CRM_Core_DAO::executeQuery($query, array(
+        1 => array($caseActivity->case_id, 'Integer')
+      ));
+
+      if ($dbResult->N === 0) {
+        return TRUE;
+      }
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * Checks if the given case activity is closing the case.
+   *
+   * @param CRM_Core_DAO $caseActivity
+   *   Object with the data for the case and activity.
+   *
+   * @return bool
+   *   True if the current activity is closing the case, false otherwise
+   */
+  private function wasCaseClosing($caseActivity) {
+    $activityType = CRM_Utils_Array::value($caseActivity->activity_type_id, $this->activityTypes);
+
+    if ($activityType === 'Change Case Status') {
+      if ($this->extractStatusFromSubject($caseActivity->subject) === 'Closed') {
+        return TRUE;
+      }
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * Given an activity's subject, it tries to extract from the new case status,
+   * if it was opened or closed.
+   *
+   * @param string $activitySubject
+   *
+   * @return string|null
+   *   Returns 'Opened' or 'Closed' if it is able to extract the case status from
+   *   the subject. Otherwise it returns null.
+   */
+  private function extractStatusFromSubject($activitySubject) {

```

Comment:
These kinds of methods are ideal for util classes and then have plenty of tests for them